relay-cardinality/src/limiter.rs

```diff
@@ -228,6 +232,8 @@ pub struct CardinalityLimitsSplit<T> {
     pub accepted: Vec<T>,
     /// The list of rejected elements of the source.
     pub rejected: Vec<T>,
+    /// The list of cardinality limits exceeded by each element of `rejected`.
+    pub exceeded_limits: Vec<SmallVec<[String; 4]>>,
```

Comment:
I think we shouldn't expose this publically. 

When building the 'result' we know for a fact that for every rejected item there is at least one reason/limit why it was rejected. So instead I'd encode it directly in the `rejected` type, e.g. `Vec<(T, Limits<'a>)>` or if for performance or other reasons storing that information separate in a second vector makes more sense, I'd not make it public but provide a method which gives you an already zipped iterator.

Reply:
Yeah, makes sense. It would require adding a lifetime parameter to `CardinalityLimitsSplit` though, is that acceptable?

Reply:
> Is that acceptable?

Yeah, I think so, the `CardinalityLimitsSplit` was just something to name what we had before as a tuple of `(Vec::new(), Vec::new())`, now we give it a bit more data, which happens to contain reference back to the limits passed in.